File: packages/control-plane-ui/src/api/services/infrastructure/ops/opsmetrics.ts
/**
 * Ops-metrics admin API service — bound to `/api/admin/ops-metrics/*`.
 *
 * The Control Plane reads `metric_ops_*` (raw + 1h/1d/1mo rollups) directly
 * via pgx; this service exposes the three CP endpoints that surface those
 * tables to the admin UI:
 *
 *   GET /api/admin/ops-metrics/current     — latest sample per (node, metric, dim)
 *                                             within the last 90s.
 *   GET /api/admin/ops-metrics/timeseries  — bucketed series for one tuple.
 *   GET /api/admin/ops-metrics/fleet       — fleet-aggregate (nodeId IS NULL)
 *                                             buckets for `nodeType=agent`.
 *
 * Wire shapes mirror the Go structs in
 * `packages/control-plane/internal/store/opsmetrics_store.go` (OpsMetricSample
 * + OpsMetricBucket); the CP handler always wraps payloads in `{ data: [...] }`,
 * with `granularity` echoed back on the timeseries / fleet endpoints so the
 * UI can show how the server resolved an `auto` request.
 */

import { api } from '../../../client';

export type NodeTypeFilter =
  | 'service'
  | 'agent'
  | 'control-plane'
  | 'ai-gateway'
  | 'compliance-proxy'
  | 'nexus-hub';

export type MetricKind = 'gauge' | 'counter' | 'histogram';

export type Granularity = 'auto' | 'raw' | '1h' | '1d' | '1mo';

/** Resolved granularity that comes back from the server (no `auto`). */
export type ResolvedGranularity = Exclude<Granularity, 'auto'>;

/** One row from `metric_ops_raw` (or its synthesized "latest per key" view). */
export interface OpsMetricSample {
  sampledAt: string;
  nodeId: string;
  nodeType: string;
  metricName: string;
  metricKind: MetricKind | string;
  dimensionKey: string;
  /** Counter increments and gauges; histograms have no scalar value. */
  value?: number | null;
  metadata?: Record<string, unknown>;
}

/** One bucket from a rollup table (or a single-row view of a raw sample). */
export interface OpsMetricBucket {
  bucketStart: string;
  /** Null for fleet aggregates (per-node rows always set nodeId). */
  nodeId?: string | null;
  nodeType: string;
  metricName: string;
  metricKind: MetricKind | string;
  dimensionKey: string;
  valueAvg?: number | null;
  valueSum?: number | null;
  valueMin?: number | null;
  valueMax?: number | null;
  sampleCount: number;
  metadata?: Record<string, unknown>;
}

export interface OpsMetricsCurrentParams {
  nodeType?: NodeTypeFilter;
  nodeId?: string;
}

export interface OpsMetricsTimeseriesParams {
  nodeId: string;
  metric: string;
  from: string;
  to: string;
  dim?: string;
  granularity?: Granularity;
}

export interface OpsMetricsFleetParams {
  nodeType: NodeTypeFilter;
  metric: string;
  from: string;
  to: string;
  dim?: string;
  granularity?: Granularity;
}

export interface TimeseriesResponse {
  data: OpsMetricBucket[];
  granularity: ResolvedGranularity;
}

/**
 * Build the `?…` suffix or empty string for a flat string-keyed param object.
 * Empty/undefined values are dropped so the URL stays canonical.
 */
function qs(p: Record<string, string | undefined>): string {
  const u = new URLSearchParams();
  for (const [k, v] of Object.entries(p)) {
    if (v != null && v !== '') u.set(k, v);
  }
  const s = u.toString();
  return s ? `?${s}` : '';
}

export const opsMetricsApi = {
  current: (p: OpsMetricsCurrentParams = {}) =>
    api
      .get<{ data: OpsMetricSample[] }>(`/api/admin/ops-metrics/current${qs({
        nodeType: p.nodeType,
        nodeId: p.nodeId,
      })}`)
      .then((r) => r.data),

  timeseries: (p: OpsMetricsTimeseriesParams) =>
    api.get<TimeseriesResponse>(
      `/api/admin/ops-metrics/timeseries${qs({
        nodeId: p.nodeId,
        metric: p.metric,
        from: p.from,
        to: p.to,
        dim: p.dim,
        granularity: p.granularity,
      })}`,
    ),

  fleet: (p: OpsMetricsFleetParams) =>
    api.get<TimeseriesResponse>(
      `/api/admin/ops-metrics/fleet${qs({
        nodeType: p.nodeType,
        metric: p.metric,
        from: p.from,
        to: p.to,
        dim: p.dim,
        granularity: p.granularity,
      })}`,
    ),
};

File: packages/control-plane/internal/observability/opsmetrics/handler/opsmetrics.go
// Package handler — opsmetrics.go: HTTP handlers for /api/admin/ops-metrics/*.
// Reads run against the metric_ops_* tables CP shares with Hub. Writes are
// limited to retention config + diag-mode windows (in separate files).
package opsmetrics

import (
	"net/http"
	"strings"
	"time"

	"github.com/labstack/echo/v4"

	"github.com/AlphaBitCore/nexus-gateway/packages/control-plane/internal/observability/opsmetrics/opsstore"
	"github.com/AlphaBitCore/nexus-gateway/packages/shared/identity/iam"
)

// RegisterOpsMetricsRoutes wires GET /api/admin/ops-metrics/{current,timeseries,fleet}.
// All three are read-only and gated by admin:ReadObservability per spec §10.
func (h *Handler) RegisterOpsMetricsRoutes(g *echo.Group, iamMW func(action string) echo.MiddlewareFunc) {
	g.GET("/ops-metrics/current", h.OpsMetricsCurrent, iamMW(iam.ResourceObservability.Action(iam.VerbRead)))
	g.GET("/ops-metrics/timeseries", h.OpsMetricsTimeseries, iamMW(iam.ResourceObservability.Action(iam.VerbRead)))
	g.GET("/ops-metrics/fleet", h.OpsMetricsFleet, iamMW(iam.ResourceObservability.Action(iam.VerbRead)))
}

// OpsMetricsCurrent returns the latest sample per (thing_id, metric_name,
// dimension_key) within the last 90s, optionally filtered by nodeType /
// nodeId.
func (h *Handler) OpsMetricsCurrent(c echo.Context) error {
	if h.ops == nil {
		return c.JSON(http.StatusServiceUnavailable, errJSON("Database is not configured", "server_error", "DB_UNAVAILABLE"))
	}
	thingType := strings.TrimSpace(c.QueryParam("nodeType"))
	if thingType != "" && thingType != "service" && thingType != "agent" &&
		thingType != "control-plane" && thingType != "ai-gateway" &&
		thingType != "compliance-proxy" && thingType != "nexus-hub" {
		return c.JSON(http.StatusBadRequest, errJSON("invalid nodeType", "validation_error", "VALIDATION_ERROR"))
	}
	thingID := strings.TrimSpace(c.QueryParam("nodeId"))

	samples, err := h.ops.GetOpsMetricsCurrent(c.Request().Context(), opsstore.OpsCurrentParams{
		ThingType: thingType,
		ThingID:   thingID,
	})
	if err != nil {
		h.logger.Error("ops_metrics_current", "error", err)
		return c.JSON(http.StatusInternalServerError, errJSON("Failed to read ops metrics", "server_error", "INTERNAL_ERROR"))
	}
	if samples == nil {
		samples = []opsstore.OpsMetricSample{}
	}
	return c.JSON(http.StatusOK, map[string]any{"data": samples})
}

// OpsMetricsTimeseries returns time-bucketed metric values for one
// (nodeId, metric, dim?) tuple. Granularity defaults to "auto".
func (h *Handler) OpsMetricsTimeseries(c echo.Context) error {
	if h.ops == nil {
		return c.JSON(http.StatusServiceUnavailable, errJSON("Database is not configured", "server_error", "DB_UNAVAILABLE"))
	}
	p, herr := parseOpsTimeseriesParams(c)
	if herr != nil {
		return c.JSON(herr.status, herr.body)
	}

	buckets, err := h.ops.GetOpsMetricsTimeseries(c.Request().Context(), p)
	if err != nil {
		h.logger.Error("ops_metrics_timeseries", "error", err, "metric", p.MetricName, "thing", p.ThingID)
		return c.JSON(http.StatusInternalServerError, errJSON("Failed to read ops timeseries", "server_error", "INTERNAL_ERROR"))
	}
	if buckets == nil {
		buckets = []opsstore.OpsMetricBucket{}
	}
	return c.JSON(http.StatusOK, map[string]any{
		"data":        buckets,
		"granularity": p.Granularity,
	})
}

// OpsMetricsFleet returns the fleet-aggregate (thing_id IS NULL) bucket series
// for the given metric over [from, to). Used by the agent fleet rollup view.
func (h *Handler) OpsMetricsFleet(c echo.Context) error {
	if h.ops == nil {
		return c.JSON(http.StatusServiceUnavailable, errJSON("Database is not configured", "server_error", "DB_UNAVAILABLE"))
	}
	thingType := strings.TrimSpace(c.QueryParam("nodeType"))
	if thingType == "" {
		return c.JSON(http.StatusBadRequest, errJSON("nodeType is required", "validation_error", "VALIDATION_ERROR"))
	}
	metric := strings.TrimSpace(c.QueryParam("metric"))
	if metric == "" {
		return c.JSON(http.StatusBadRequest, errJSON("metric is required", "validation_error", "VALIDATION_ERROR"))
	}
	from, to, herr := parseFromTo(c)
	if herr != nil {
		return c.JSON(herr.status, herr.body)
	}
	gran := strings.TrimSpace(c.QueryParam("granularity"))
	if gran == "" || gran == "auto" {
		gran = opsstore.SelectGranularity(from, to)
		if gran == "raw" {
			gran = "5m" // raw has no fleet aggregate; smallest fleet tier is 5m
		}
	}
	if gran != "5m" && gran != "1h" && gran != "1d" && gran != "1mo" {
		return c.JSON(http.StatusBadRequest, errJSON("invalid granularity for fleet (need 5m|1h|1d|1mo)", "validation_error", "VALIDATION_ERROR"))
	}

	var dim *string
	if v, ok := c.QueryParams()["dim"]; ok && len(v) > 0 {
		s := v[0]
		dim = &s
	}

	buckets, err := h.ops.GetOpsMetricsFleet(c.Request().Context(), opsstore.OpsFleetParams{
		ThingType:    thingType,
		MetricName:   metric,
		DimensionKey: dim,
		From:         from,
		To:           to,
		Granularity:  gran,
	})
	if err != nil {
		h.logger.Error("ops_metrics_fleet", "error", err, "metric", metric, "nodeType", thingType)
		return c.JSON(http.StatusInternalServerError, errJSON("Failed to read fleet metrics", "server_error", "INTERNAL_ERROR"))
	}
	if buckets == nil {
		buckets = []opsstore.OpsMetricBucket{}
	}
	return c.JSON(http.StatusOK, map[string]any{
		"data":        buckets,
		"granularity": gran,
	})
}

// Param parsing helpers (also used by diag handlers)

// httpErr packages an HTTP status + JSON body so handlers can short-circuit
// from a helper without juggling two return values per failure case.
type httpErr struct {
	status int
	body   map[string]any
}

func badReq(msg string) *httpErr {
	return &httpErr{
		status: http.StatusBadRequest,
		body:   errJSON(msg, "validation_error", "VALIDATION_ERROR"),
	}
}

// parseFromTo extracts ?from=&to= as RFC3339 timestamps. Both are required.
func parseFromTo(c echo.Context) (time.Time, time.Time, *httpErr) {
	fromStr := strings.TrimSpace(c.QueryParam("from"))
	toStr := strings.TrimSpace(c.QueryParam("to"))
	if fromStr == "" || toStr == "" {
		return time.Time{}, time.Time{}, badReq("from and to are required (RFC3339)")
	}
	from, ok := parseRFC3339Flexible(fromStr)
	if !ok {
		return time.Time{}, time.Time{}, badReq("invalid from (need RFC3339)")
	}
	to, ok2 := parseRFC3339Flexible(toStr)
	if !ok2 {
		return time.Time{}, time.Time{}, badReq("invalid to (need RFC3339)")
	}
	if !from.Before(to) {
		return time.Time{}, time.Time{}, badReq("from must be < to")
	}
	return from, to, nil
}

// parseOpsTimeseriesParams normalises ?nodeId=&metric=&dim=&from=&to=&granularity=.
// Granularity defaults to auto-pick on the from/to span.
func parseOpsTimeseriesParams(c echo.Context) (opsstore.OpsTimeseriesParams, *httpErr) {
	thingID := strings.TrimSpace(c.QueryParam("nodeId"))
	if thingID == "" {
		return opsstore.OpsTimeseriesParams{}, badReq("nodeId is required")
	}
	metric := strings.TrimSpace(c.QueryParam("metric"))
	if metric == "" {
		return opsstore.OpsTimeseriesParams{}, badReq("metric is required")
	}
	from, to, herr := parseFromTo(c)
	if herr != nil {
		return opsstore.OpsTimeseriesParams{}, herr
	}

	gran := strings.TrimSpace(c.QueryParam("granularity"))
	if gran == "" || gran == "auto" {
		gran = opsstore.SelectGranularity(from, to)
	}
	switch gran {
	case "raw", "5m", "1h", "1d", "1mo":
	default:
		return opsstore.OpsTimeseriesParams{}, badReq("invalid granularity (need auto|raw|5m|1h|1d|1mo)")
	}

	var dim *string
	if v, ok := c.QueryParams()["dim"]; ok && len(v) > 0 {
		s := v[0]
		dim = &s
	}

	return opsstore.OpsTimeseriesParams{
		ThingID:      thingID,
		MetricName:   metric,
		DimensionKey: dim,
		From:         from,
		To:           to,
		Granularity:  gran,
	}, nil
}
